lib/rdoc/parser/prism_ruby.rb

```diff
@@ -0,0 +1,1081 @@
+        receiver_fallback_type: receiver_fallback_type,
+        visibility: visibility,
+        singleton: singleton,
+        signature: signature,
+        tokens: tokens,
+        start_line: start_line,
+        end_line: end_line
+      )
+    ensure
+      @scanner.skip_comments_until(end_line)
+    end
+
+    def visit_constant_path_write_node(node)
+      @scanner.process_comments_until(node.location.start_line - 1)
+      path = constant_path_string(node.target)
+      return unless path
+
+      @scanner.add_constant(
+        path,
+        constant_path_string(node.value) || node.value.slice,
+        node.location.start_line,
+        node.location.end_line
+      )
+      @scanner.skip_comments_until(node.location.end_line)
+      # Do not traverse rhs not to document `A::B = Struct.new{def undocumentable_method; end}`
+    end
+
+    def visit_constant_write_node(node)
+      @scanner.process_comments_until(node.location.start_line - 1)
+      @scanner.add_constant(
+        node.name.to_s,
+        constant_path_string(node.value) || node.value.slice,
+        node.location.start_line,
+        node.location.end_line
+      )
+      @scanner.skip_comments_until(node.location.end_line)
+      # Do not traverse rhs not to document `A = Struct.new{def undocumentable_method; end}`
+    end
+
+    private
+
+    def constant_arguments_names(call_node)
+      return unless call_node.arguments
+      names = call_node.arguments.arguments.map { |arg| constant_path_string(arg) }
+      names.all? ? names : nil
+    end
+
+    def symbol_arguments(call_node)
+      arguments_node = call_node.arguments
+      return unless arguments_node && arguments_node.arguments.all? { |arg| arg.is_a?(Prism::SymbolNode)}
+      arguments_node.arguments.map { |arg| arg.value.to_sym }
+    end
+
+    def visibility_method_arguments(call_node, singleton:)
+      arguments_node = call_node.arguments
+      return unless arguments_node
+      symbols = symbol_arguments(call_node)
+      return symbols.map(&:to_s) if symbols # module_function :foo, :bar
+      return unless arguments_node.arguments.size == 1
+      arg = arguments_node.arguments.first
+      return unless arg.is_a?(Prism::DefNode)
+      # `module_function def foo; end` or `private_class_method def self.foo; end`
+      [arg.name.to_s] if singleton ? arg.receiver.is_a?(Prism::SelfNode) : arg.receiver.nil?
+    end
+
+    def constant_path_string(node)
+      case node
+      when Prism::ConstantReadNode
+        node.name.to_s
+      when Prism::ConstantPathNode
+        parent_name = node.parent ? constant_path_string(node.parent) : ''
+        "#{parent_name}::#{node.name}" if parent_name
+      end
+    end
+  end
+
+  class MethodSignature < Prism::Visitor # :nodoc:
+    attr_reader :yields, :calls_super, :params
+    def initialize(def_node)
+      @yields = []
+      @calls_super = false
+      @params = "(#{def_node.parameters&.slice})"
+      def_node.body&.accept(self)
```

Comment:
I think for readability visitors should NOT invoke itself, so we should move this out of `initialize` and do this in `visit_def_node` instead

```rb
      if node.body
        signature = MethodSignature.new(node)
        node.body.accept(signature)
      end
```